### PR TITLE
Add Chapter 4 exercise 2 ReversingAnArray KAN-24 

### PR DESCRIPTION
**Pull request by insert-oreyes**
- Add component ReversingAnArray with the creative result, documentation and testing
- Update Util component with the functions to be used in ReversingAnArray, documentation and testing
- Update db with exercise 2 details
- Update Chapter 4 component rendering exercise 2 and testing

https://github.com/jon-oscar/eloquent-js/assets/99765175/e098e06f-c6c1-4446-99fb-ec222ce45e96

### CODE REVIEW

File: components/exercises/chapter-4-oscar/Utils.tsx [hunk 1]
```diff
@@ -7,34 +7,67 @@
...
+  for (let i = 0; i < arr.length; i++) sum += arr[i];
+  return sum;
+};
+
+/**
+ * Reverses the order of elements in an array.
+ * @param arr - The array to be reversed.
+ * @returns A new array with the elements in reverse order.
+ */
+export function getReverseArray(arr: any[]): any[] {
```
Comment (jportella93) [inline components/exercises/chapter-4-oscar/Utils.tsx]: better naming: `getReversedArray`

File: components/exercises/chapter-4-oscar/Utils.tsx [hunk 2]
```diff
@@ -7,34 +7,67 @@
...
+  const newArr = [];
+
+  // loop through the array backwards and arr.length - 1 = 4 to match with indexation
```
Comment (jportella93) [inline components/exercises/chapter-4-oscar/Utils.tsx]: This comment doesn't make sense to me, better description could be:  
```js
// loop through the passed array in inverse order and add each item to the new array
```

File: components/exercises/chapter-4-oscar/Utils.test.tsx [hunk 3]
```diff
@@ -30,3 +35,40 @@ describe('arraySum', () => {
...
+});
+
+describe('getReverseArrayInPlace', () => {
+  it('should return an empty array if given an empty array', () => {
+    const arr: any[] = [];
+    getReverseArrayInPlace(arr);
+    expect(arr).toEqual([]);
+  });
+
+  it('should return an array with the elements in reverse order', () => {
```
Comment (jportella93) [inline components/exercises/chapter-4-oscar/Utils.test.tsx:73]: can we add a test for an array with length different than 5?

File: components/exercises/chapter-4-oscar/Utils.tsx [hunk 4]
```diff
@@ -7,34 +7,67 @@
...
+
+    // array[0] is assigned the value at array[4], which is 5
+    arr[i] = arr[arr.length - 1 - i];
+
+    // array[4] is assigned the value of prevArr, which is 1
+    arr[arr.length - 1 - i] = prevArr;
+
+    // After the first iteration, the array becomes [5, 2, 3, 4, 1]
+  }
+  return arr;
```
Comment (jportella93) [inline components/exercises/chapter-4-oscar/Utils.tsx]: This function **mutates** the original array so:
- You don't need to return anything at the end of the function
- You can rename the function to `reverseArrayInPlace`

I suggest you read this https://www.phind.com/search?cache=n6825r14yaakvmhyzj61toh0

File: components/exercises/chapter-4-oscar/Utils.test.tsx [hunk 5]
```diff
@@ -30,3 +35,40 @@ describe('arraySum', () => {
     expect(arraySum([-1, -2, -3, -4, -5])).toEqual(-15);
   });
 });
+
+describe('getReverseArray', () => {
```
Comment (jportella93) [inline components/exercises/chapter-4-oscar/Utils.test.tsx]: Can we add a test to make sure the array returned is not the same array as the one passed? (so it is a copy)

**Review by jportella93 — COMMENTED** [review]
Added some comments. Read up on array mutation on JavaScript and make sure you understand it, it is one of those mid-advanced topics that makes the difference between a basic and a pro engineer

File: components/exercises/chapter-4-oscar/Utils.tsx (hunk 1 above)
Comment (insert-oreyes) [reply components/exercises/chapter-4-oscar/Utils.tsx]: refactored

File: components/exercises/chapter-4-oscar/Utils.tsx (hunk 2 above)
Comment (insert-oreyes) [reply components/exercises/chapter-4-oscar/Utils.tsx]: refactored

File: components/exercises/chapter-4-oscar/Utils.test.tsx (hunk 3 above)
Comment (insert-oreyes) [reply components/exercises/chapter-4-oscar/Utils.test.tsx:73]: refactor adding length of 3 and length of 10

File: components/exercises/chapter-4-oscar/Utils.tsx (hunk 4 above)
Comment (insert-oreyes) [reply components/exercises/chapter-4-oscar/Utils.tsx]: - I am getting the following error: "A function whose declared type is neither 'undefined, 'void', nor 'any
 must return a value".
According to what I found it must return something or enable --noImplicitReturns
https://github.com/microsoft/TypeScript/issues/53559
- Refactored

File: components/exercises/chapter-4-oscar/Utils.test.tsx (hunk 5 above)
Comment (insert-oreyes) [reply components/exercises/chapter-4-oscar/Utils.test.tsx]: test add and passed

File: components/exercises/chapter-4-oscar/Utils.tsx (hunk 4 above)
Comment (jportella93) [reply components/exercises/chapter-4-oscar/Utils.tsx]: when a function doesn't return anything, it returns implicitly `undefined`, so to fix the typescript error, you should update the typing to return `undefined`

**Review by jportella93 — APPROVED** [review]

File: components/exercises/chapter-4-oscar/Utils.tsx (hunk 4 above)
Comment (insert-oreyes) [reply components/exercises/chapter-4-oscar/Utils.tsx]: fixed, thank you

**Pull request merged**